### PR TITLE
Auto generate font

### PR DESCRIPTION
**Pull request by yehoshuapw**
closes #964 (supersedes)

this adds a script `src/displayapp/fonts/generate.py` to generate the font files (using lv_font_conv) from a singular config file (`src/displayapp/fonts/fonts.json`).

the json has each font, and it's relevant options (size,bpp,compress, and "patches") and sources.
in addition there's also a "feature" feature, allowing extra sources with a extra flag (and the config has an example usage of adding hebrew, which won't do anything unless `-e hebrew` is given)

currently, this manages to generate exactly the same as what is already in place.

TODO:
integrate this into the CMake (and add lv_font_conv to the build images).
(my CMake-fu is not that strong, so I'll get to that tomorrow - but opening this to get comments about the rest of it from anyone interested)

### CODE REVIEW

**Comment by yehoshuapw** [comment]
@Riksu9000 i'd like to get your thoughts about whats already here. thanks!

File: src/displayapp/fonts/generate.py
Comment (Riksu9000) [inline src/displayapp/fonts/generate.py:61]: I think there should be a warning when the font that has been entered as an option can't be found. This could happen just as a typo. Perhaps this could loop through the fonts_to_run instead of all fonts in the config.

File: src/displayapp/fonts/generate.py
Comment (Riksu9000) [inline src/displayapp/fonts/generate.py:57]: This seems useful, but I think this could be automated and the argument removed.

File: src/displayapp/fonts/fix_jetbrains_mono_bold_20_zero.py
Comment (Riksu9000) [inline src/displayapp/fonts/fix_jetbrains_mono_bold_20_zero.py:1]: Maybe we could use a normal patch file for patches.

File: src/displayapp/fonts/fonts.json
Comment (Riksu9000) [inline src/displayapp/fonts/fonts.json:89]: Features should probably be inside the corresponding font section

**Review by Riksu9000 — COMMENTED** [review]
I like this solution. The README should be updated as a part of this PR by removing the old instructions and writing something about this new method.

File: src/displayapp/fonts/generate.py
Comment (yehoshuapw) [reply src/displayapp/fonts/generate.py:67]: Added a check (which is only a warning, not an error).
not looping only thorough "fonts_to_run" since by default (without `-f`s) will run over all the fonts in the config file.

File: src/displayapp/fonts/generate.py
Comment (yehoshuapw) [reply src/displayapp/fonts/generate.py:57]: Indeed - and this is exactly such an automation. do you think it will be more clear if done inside the CMake files, and not here? (I added it in it's own commit after deciding it will make the CMake more readable - if you disagree, I'll remove it)

File: src/displayapp/fonts/fix_jetbrains_mono_bold_20_zero.py
Comment (yehoshuapw) [reply src/displayapp/fonts/fix_jetbrains_mono_bold_20_zero.py:1]: this will work only so long as the relevant lines stay in the same place, and look exactly the same (including the comment, which just changed between versions of lv_font_conv), which is why I made this over-complicated patcher script.

(I almost made it even more of a mess, by allowing for space changes and such).

so I recommend to leave it like this, but this really is a mess - and I'll abide by your decision.

File: src/displayapp/fonts/fonts.json
Comment (yehoshuapw) [reply src/displayapp/fonts/fonts.json:89]: that will simplify the script, but will make the config less readable. like this the "features" can be clearly ignored by a human reading the config.

Comment (yehoshuapw) [reply src/displayapp/fonts/fonts.json]: Added commit which makes this change.

File: src/displayapp/fonts/generate.py
Comment (yehoshuapw) [reply src/displayapp/fonts/generate.py]: done

File: src/displayapp/fonts/generate.py
Comment (Riksu9000) [reply src/displayapp/fonts/generate.py:55]: `fonts_to_run` could also by default contain all the fonts in the config file.

File: src/displayapp/fonts/fix_jetbrains_mono_bold_20_zero.py
Comment (yehoshuapw) [reply src/displayapp/fonts/fix_jetbrains_mono_bold_20_zero.py:1]: nvm about the comment (the diff does not need to include parts before and after), but the lines do still matter - which can change anytime a new symbol is added which ends up before.

File: src/displayapp/fonts/generate.py
Comment (Riksu9000) [reply src/displayapp/fonts/generate.py]: I meant it more like the script could automatically strip the extension, but I guess we can decide what to do when we have some CMake code to review. I see it as a possibiliity for this to be merged even without any CMake integration initially depending on the situation.

Comment (yehoshuapw) [reply src/displayapp/fonts/generate.py]: changed to that method (strip .c if it's there)

File: src/displayapp/fonts/fonts.json
Comment (yehoshuapw) [reply src/displayapp/fonts/fonts.json]: it won't happen often, since most requests for font changes are either for a new language (which should be added in the same place as the hebrew example),
or for new watchfaces - which also use at most 1 font, and often a completely new one, and that need it (so it's not a toggle-able feature)

File: src/displayapp/fonts/README.md
Comment (yehoshuapw) [inline src/displayapp/fonts/README.md:10]: i'm not sure if adding npm instructions is necessary - or just annoying.

**Comment by yehoshuapw** [comment]
and now the font .c files are generated at compile time with CMake.

**Comment by Riksu9000** [comment]
I've been thinking that the font "features" probably isn't going to be very useful to us, since we only ship a single firmware, and that's likely not going to be changing. The other use case would be if fork maintainers pushed their changes in this repo, which I don't think should be our responsibility to handle either.

**Review by JF002 — CHANGES_REQUESTED** [review]
Thank you for this PR, that's definitely something I want to add tot he project!

The generation of the fonts in CMake and then in Python and in Github Actions look good to me!

If you don't mind, I think we should make 2 addition:
 - Add  `lv_font_conv` (and how to install it via `npm`) as a needed dependency to build the project in `doc/buildAndProgram.md` as the project won't build without it.
 - Add the installation of `npm` and `lv_font_conv` in the Dockerfile so we can build the fonts inside the docker container. Note than I plan on switching the CI to docker so that we won't have to duplicate those changes in the future (https://github.com/InfiniTimeOrg/InfiniTime/pull/1045).
 - Oh and also maybe specify which version of `lv_font_conv` we support, and enforce this version in Docker/Action workflows to avoid broken builds because of a change in the tool ?

I can help with those points if you want to ;)

File: src/displayapp/fonts/README.md
Comment (Riksu9000) [reply src/displayapp/fonts/README.md:10]: I think this entire section can be removed, because you're not intended to run `generate.py` manually. Since `lv_font_conv` is now a build dependency, it should just be added [here](https://github.com/InfiniTimeOrg/InfiniTime/blob/develop/doc/buildAndProgram.md) with a link to the [repo](https://github.com/lvgl/lv_font_conv) with instructions being there.

**Review by Riksu9000 — COMMENTED** [review]
I noticed that changing the font configuration will regenerate all the fonts, even if only one was changed. Not a big issue, but maybe worth considering if something can be done about that.

**Comment by yehoshuapw** [comment]
> I've been thinking that the font "features" probably isn't going to be very useful to us, since we only ship a single firmware, and that's likely not going to be changing. The other use case would be if fork maintainers pushed their changes in this repo, which I don't think should be our responsibility to handle either.

since extra languages is a commonly requested feature, I decided this will make things easier (until a real solution is implemented), but I see your point.
How about this: adding a flag to add "features" from a separate json file? then the script is aware of such things, but the main files don't need to be changed.

**Comment by yehoshuapw** [comment]
> I noticed that changing the font configuration will regenerate all the fonts, even if only one was changed. Not a big issue, but maybe worth considering if something can be done about that.

the options as I see them are:

* a config file per font
* some method to non regenerate if some part was not changed
* a single config file, and regenerate all the fonts.

since the generation is pretty fast, and changing the config does not happen often - I don't like option 1.
option 2 is doable, but will make it more complicated - and may not get much of an increase in speed.

therefore I just left it at a plain single-file. do you think one of the others is better? or do you have a better idea?
thanks!

File: src/displayapp/fonts/README.md
Comment (Riksu9000) [inline src/displayapp/fonts/README.md:12]: I think this entire section can be removed now that the dependency note has been added elsewhere and generate.py shouldn't be run manually.

**Comment by Riksu9000** [comment]
> > I've been thinking that the font "features" probably isn't going to be very useful to us, since we only ship a single firmware, and that's likely not going to be changing. The other use case would be if fork maintainers pushed their changes in this repo, which I don't think should be our responsibility to handle either.
> 
> since extra languages is a commonly requested feature, I decided this will make things easier (until a real solution is implemented), but I see your point. How about this: adding a flag to add "features" from a separate json file? then the script is aware of such things, but the main files don't need to be changed.

It's a neat feature, but adding the extra characters and keeping it up to date with git manually isn't difficult either, and it's already much easier than before since the font is generated automatically. I don't think we'd want to maintain a feature that we don't use.

**Comment by yehoshuapw** [comment]
does anyone know if `libssl-dev` as opposed to `libssl1.0-dev` matters in the docker?

File: src/displayapp/fonts/fonts.json
Comment (Riksu9000) [inline src/displayapp/fonts/fonts.json:15]: I think supplying just the name of the patch file should be enough here. It doesn't need to be able to run any command.

File: src/displayapp/fonts/README.md
Comment (Riksu9000) [inline src/displayapp/fonts/README.md:11]: There's an extra asterisk.

```suggestion
* Convert this hex value into a UTF-8 code
```

**Review by Riksu9000 — COMMENTED** [review]
I think the script should check that `lv_font_conv` is installed before running and show an error if it isn't, because otherwise it will show many errors, which may be confusing.

File: src/displayapp/fonts/fonts.json
Comment (yehoshuapw) [reply src/displayapp/fonts/fonts.json]: done. (I left in that if the script fails using a simple `patch`, it will try using as a whole command)

**Comment by yehoshuapw** [comment]
> because otherwise it will show many errors, which may be confusing.

now it will check and give a clear error message, however it will still display the same message for each - CMake runs each font by itself.

File: src/displayapp/fonts/fonts.json
Comment (Riksu9000) [reply src/displayapp/fonts/fonts.json]: I don't think the `{file}` overriding is useful for filenames here, and I'm not sure falling back to executing as a command is necessary. I'd rather remove features that aren't necessary since they won't get maintained.

Comment (Riksu9000) [reply src/displayapp/fonts/fonts.json]: Good question. As far as I can tell, that capability just requires one extra for loop, so it's not that bad. If a font required two changes, my immediate thought would be that they should be two separate patches, but I'm not 100% sure.

Comment (yehoshuapw) [reply src/displayapp/fonts/fonts.json]: then I'll leave that in.

**Review by Riksu9000 — APPROVED** [review]
For InfiniSim to compile, generate.py currently must be run manually. I think that may be something that needs to be done on InfiniSim's side. Ping @NeroBurner.

I can't comment much on the CMake changes, since I'm not so familiar with that.

Other than that I really like this and it works well.

**Comment by yehoshuapw** [comment]
> For InfiniSim

Added a bit of changes here (use CMAKE_CURRENT_LIST_DIR to reach things, instead of assuming paths),
and included this CMakeLists.txt there (and changed the GLOB there to use the list inside the CMake here, since globbing on filenames won't work anymore)

**Comment by yehoshuapw** [comment]
> > For InfiniSim

Also, I did (except when I forgot once) compile before pushing, and here watching the checks status, so it's weird that the simulator check succeeded.

so I had a look, and in the simulator workflow here I manually generated the scripts - I just removed it, so it will fail here too (until InfiniTimeOrg/InfiniSim#28 is merged).
(so both these PRs should be merged about the same time)

File: src/displayapp/fonts/CMakeLists.txt
Comment (JF002) [inline src/displayapp/fonts/CMakeLists.txt:14]:
```suggestion
      COMMAND python3 ${CMAKE_CURRENT_LIST_DIR}/generate.py
```

**Review by JF002 — CHANGES_REQUESTED** [review]
This looks totally fine to me! 
There's just 1 last issue : the build in the docker container does not work properly. See this log message:

```
[ 35%] Generating displayapp/fonts/jetbrains_mono_42.c
[ 35%] Generating displayapp/fonts/jetbrains_mono_76.c
[ 35%] Generating displayapp/fonts/jetbrains_mono_76.c
[ 35%] Generating displayapp/fonts/lv_font_navi_80.c
[ 35%] Generating displayapp/fonts/open_sans_light.c
[ 35%] Generating displayapp/fonts/open_sans_light.c
[ 35%] Generating displayapp/fonts/jetbrains_mono_extrabold_compressed.c
[ 35%] Generating displayapp/fonts/jetbrains_mono_extrabold_compressed.c
[ 35%] Generating displayapp/fonts/jetbrains_mono_bold_20.c
[ 35%] Generating displayapp/fonts/jetbrains_mono_42.c
[ 35%] Generating displayapp/fonts/lv_font_sys_48.c
[ 36%] Generating displayapp/fonts/jetbrains_mono_bold_20.c
[ 36%] Generating displayapp/fonts/lv_font_navi_80.c
  File "  File "/sources/src/displayapp/fonts/generate.py/sources/src/displayapp/fonts/generate.py", line ", line 21
21
    def gen_lvconv_line(dest: str, size: int, bpp: int, sources: typing.List[Source], compress:bool=False):
  File "/sources/src/displayapp/fonts/generate.py  File "/sources/src/displayapp/fonts/generate.py", line 21
  File "        def gen_lvconv_line(dest: str, size: int, bpp: int, sources: typing.List[Source], compress:bool=False):
                              ", line 21
    def gen_lvconv_line(dest: str, size: int, bpp: int, sources: typing.List[Source], compress:bool=False):
/sources/src/displayapp/fonts/generate.py      def gen_lvconv_line(dest: str, size: int, bpp: int, sources: typing.List[Source], compress:bool=False):
         File "  File "/sources/src/displayapp/fonts/generate.py", line  /sources/src/displayapp/fonts/generate.py21", line 
     21 def gen_lvconv_line(dest: str, size: int, bpp: int, sources: typing.List[Source], compress:bool=False):

           File "/sources/src/displayapp/fonts/generate.py", line ", line 21
 21    def gen_lvconv_line(dest: str, size: int, bpp: int, sources: typing.List[Source], compress:bool=False):

                                             def gen_lvconv_line(dest: str, size: int, bpp: int, sources: typing.List[Source], compress:bool=False):
                                ^
  SyntaxError :   invalid syntax 
     ^
SyntaxError                       ^
    SyntaxError        ^
     File "SyntaxError :  invalid syntax/sources/src/displayapp/fonts/generate.py
 ", line  21 
     def gen_lvconv_line(dest: str, size: int, bpp: int, sources: typing.List[Source], compress:bool=False):
      :   def gen_lvconv_line(dest: str, size: int, bpp: int, sources: typing.List[Source], compress:bool=False):
invalid syntax  
:  invalid syntax
          ^
    ^
SyntaxError^
: SyntaxErrorinvalid syntaxSyntaxError: 
:   File "invalid syntaxinvalid syntax
/sources/src/displayapp/fonts/generate.py
", line 21
    def gen_lvconv_line(dest: str, size: int, bpp: int, sources: typing.List[Source], compress:bool=False):
                      File "   /sources/src/displayapp/fonts/generate.py ", line   21 
     ^
def gen_lvconv_line(dest: str, size: int, bpp: int, sources: typing.List[Source], compress:bool=False):
                            ^
SyntaxError: invalid syntax
  File "/sources/src/displayapp/fonts/generate.py", line 21
    def gen_lvconv_line(dest: str, size: int, bpp: int, sources: typing.List[Source], compress:bool=False):
                            ^
SyntaxError: invalid syntax
                   ^
SyntaxError: invalid syntax
src/CMakeFiles/pinetime-mcuboot-app.dir/build.make:65: recipe for target 'src/displayapp/fonts/jetbrains_mono_extrabold_compressed.c' failed
make[2]: *** [src/displayapp/fonts/jetbrains_mono_extrabold_compressed.c] Error 1
make[2]: *** Waiting for unfinished jobs....
    src/CMakeFiles/pinetime-app.dir/build.make:73: recipe for target 'src/displayapp/fonts/jetbrains_mono_76.c' failed
 SyntaxErrormake[2]: *** [src/displayapp/fonts/jetbrains_mono_76.c] Error 1
    :   File " [ 36%] Generating displayapp/fonts/lv_font_sys_48.c
invalid syntaxmake[2]: *** Waiting for unfinished jobs....
/sources/src/displayapp/fonts/generate.py 
src/CMakeFiles/pinetime-mcuboot-app.dir/build.make:73: recipe for target 'src/displayapp/fonts/jetbrains_mono_76.c' failed
make[2]: *** [src/displayapp/fonts/jetbrains_mono_76.c] Error 1
", line 21 
    def gen_lvconv_line(dest: str, size: int, bpp: int, sources: typing.List[Source], compress:bool=False):
    src/CMakeFiles/pinetime-app.dir/build.make:77: recipe for target 'src/displayapp/fonts/jetbrains_mono_42.c' failed
          make[2]: *** [src/displayapp/fonts/jetbrains_mono_42.c] Error 1
               ^
    SyntaxError :   invalid syntax 
       ^
SyntaxError: invalid syntax
src/CMakeFiles/pinetime-app.dir/build.make:65: recipe for target 'src/displayapp/fonts/jetbrains_mono_extrabold_compressed.c' failed
make[2]: *** [src/displayapp/fonts/jetbrains_mono_extrabold_compressed.c] Error 1
src/CMakeFiles/pinetime-mcuboot-app.dir/build.make:77: recipe for target 'src/displayapp/fonts/jetbrains_mono_42.c' failed
make[2]: *** [src/displayapp/fonts/jetbrains_mono_42.c] Error 1
src/CMakeFiles/pinetime-mcuboot-app.dir/build.make:61: recipe for target 'src/displayapp/fonts/lv_font_navi_80.c' failed
make[2]: *** [src/displayapp/fonts/lv_font_navi_80.c] Error 1
src/CMakeFiles/pinetime-app.dir/build.make:69: recipe for target 'src/displayapp/fonts/jetbrains_mono_bold_20.c' failed
make[2]: *** [src/displayapp/fonts/jetbrains_mono_bold_20.c] Error 1
src/CMakeFiles/pinetime-app.dir/build.make:85: recipe for target 'src/displayapp/fonts/open_sans_light.c' failed
make[2]: *** [src/displayapp/fonts/open_sans_light.c] Error 1
src/CMakeFiles/pinetime-app.dir/build.make:81: recipe for target 'src/displayapp/fonts/lv_font_sys_48.c' failed
make[2]: *** [src/displayapp/fonts/lv_font_sys_48.c] Error 1
src/CMakeFiles/pinetime-mcuboot-app.dir/build.make:85: recipe for target 'src/displayapp/fonts/open_sans_light.c' failed
make[2]: *** [src/displayapp/fonts/open_sans_light.c] Error 1
src/CMakeFiles/pinetime-mcuboot-app.dir/build.make:69: recipe for target 'src/displayapp/fonts/jetbrains_mono_bold_20.c' failed
make[2]: *** [src/displayapp/fonts/jetbrains_mono_bold_20.c] Error 1
src/CMakeFiles/pinetime-app.dir/build.make:61: recipe for target 'src/displayapp/fonts/lv_font_navi_80.c' failed
make[2]: *** [src/displayapp/fonts/lv_font_navi_80.c] Error 1
CMakeFiles/Makefile2:409: recipe for target 'src/CMakeFiles/pinetime-app.dir/all' failed
make[1]: *** [src/CMakeFiles/pinetime-app.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
  File "/sources/src/displayapp/fonts/generate.py", line 21
    def gen_lvconv_line(dest: str, size: int, bpp: int, sources: typing.List[Source], compress:bool=False):
                            ^
SyntaxError: invalid syntax
src/CMakeFiles/pinetime-mcuboot-app.dir/build.make:81: recipe for target 'src/displayapp/fonts/lv_font_sys_48.c' failed
make[2]: *** [src/displayapp/fonts/lv_font_sys_48.c] Error 1
CMakeFiles/Makefile2:214: recipe for target 'src/CMakeFiles/pinetime-mcuboot-app.dir/all' failed
make[1]: *** [src/CMakeFiles/pinetime-mcuboot-app.dir/all] Error 2
```

I think this is caused by the version of Python : the docker image uses Python 2.7.17, while my computer is running Python 3.10.

Do you think it would cause any issue if we called `python3` instead of `python` in the cmake command that generates the font?

File: src/displayapp/fonts/CMakeLists.txt
Comment (yehoshuapw) [reply src/displayapp/fonts/CMakeLists.txt:14]: the typing should be solved by using python3 instead of 2, but this will break some newer linux builds - which don't have a "python3" executable anymore.
Also, in old enough distributions perhaps the python3 is not new enough to have f-strings,

so Instead I'll convert things to version agnostic code.

Comment (yehoshuapw) [reply src/displayapp/fonts/CMakeLists.txt]: that worked, but ended up being a bit more of a mess then expected,
so I changed to use python3, and when it's not found to use python. (if there's only python2, it will find it - and fail when running.)

Comment (JF002) [reply src/displayapp/fonts/CMakeLists.txt]: Or maybe we should update the version of the base image of the docker (to a version that uses python3 by default)? 18.04 is probably quite old.

Comment (yehoshuapw) [reply src/displayapp/fonts/CMakeLists.txt:14]: changed to use ubuntu 20.04 (the current LTS), which does indeed solve the issue without any of the other parts (which I removed)

**Comment by JF002** [comment]
@yehoshuapw I've just tried the new docker image based on ubunutu 20.04, and I still got the same issues with the Python script. 
It looks like the container based on Ubuntu 20.04 is still based on Python 2.7:

 ```
infinitime@cacd01d36e16:/$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 20.04.4 LTS
Release:        20.04
Codename:       focal
infinitime@cacd01d36e16:/$ python --version
Python 2.7.18
infinitime@cacd01d36e16:/$ python3 --version
Python 3.8.10
```
I'm not sure why it works for you and not for me... Did I miss anything?

PS : Please excuse me for the time it takes to finish this review... Those little details about Docker and the CI are taking more time than I expected...

**Comment by JF002** [comment]
> Ubuntu 22.04 is released and it has a package called `python-is-python3` which symlinks python to python3. Would this be acceptable?

Yes, probably! As long as it still builds the firmware and generates the DFU images, that fine for me :)

**Comment by yehoshuapw** [comment]
> I'm not sure why it works for you and not for me... Did I miss anything?

I must have made some mistake (left the python2 version? left .c files in the build dir?), because indeed the image I built here does have the "python" version as 2.7.

updated as you suggested, and seems to work. (building the image and then infinitime, from the start)

(I'm also wondering if any dist except for arch (which I don't even use) doesn't have a "python3" anymore without installing some extra things. nonetheless having a newer version which works tends to be worth it)

**Comment by JF002** [comment]
> (I'm also wondering if any dist except for arch (which I don't even use) doesn't have a "python3" anymore without installing some extra things. nonetheless having a newer version which works tends to be worth it)

On Manjaro, I still have python and python3, but since python 3 is support to replace python 2, it's probably better to assume that python is Python 3

> updated as you suggested, and seems to work. (building the image and then infinitime, from the start)

Thanks, that looks very good to me :)

**Review by JF002 — APPROVED** [review]

**Comment by JF002** [comment]
Now... How do we handle that in InfiniSim ?

**Comment by yehoshuapw** [comment]
> Now... How do we handle that in InfiniSim ?

I opened a PR there too, https://github.com/InfiniTimeOrg/InfiniSim/pull/28 which seems to work.

**Comment by NeroBurner** [comment]
I think nearly identical to here. Need to install npm and the lv-font generator in the CI. And need to add those generated fonts to InfiniSim. I'll need to have a thorough look at this PR and see how and where to fonts.c files are generated.
For my personal guideline, every generated file should be in the `${CMAKE_BINARY_DIR}` and not in the source directory.

Maybe I can make the font-generating CMakeLists.txt file generic enough so I just need to do `add_subdirectory` in InfiniSim and link to a infinisim-fonts target or something like that :thinking:

**Comment by yehoshuapw** [comment]
> where to fonts.c files are generated. 
in `build/displayapp/fonts` (or wherever the build dir is)

> Maybe I can make the font-generating CMakeLists.txt file generic enough so I just need to do `add_subdirectory` in InfiniSim and link to a infinisim-fonts target or something like that thinking

I'm new at CMake, so go for it.

**Comment by JF002** [comment]
I'll merge this PR as it's working as expected in InfiniTime. We'll follow up with the integration in InfiniSim in anothe PR (probably in the InfiniSim project).

> Maybe I can make the font-generating CMakeLists.txt file generic enough so I just need to do add_subdirectory in InfiniSim and link to a infinisim-fonts target or something like that thinking

It should be possible to create a 'font' target or a GenerateFont() function that we'll be able to include in InfiniSim, so we don't have to duplicate the whole code needed to generate the fonts.

@yehoshuapw I'm happy to see you've already started work in InfiniSim too! Thanks again for this nice PR! :1st_place_medal:

**Pull request merged**
